include/fiction/algorithms/physical_design/design_sidb_gates.hpp

```diff
@@ -0,0 +1,291 @@
+        std::iota(numbers.begin(), numbers.end(), 0);
+
+        combinations::for_each_combination(
+            numbers.begin(),
+            numbers.begin() + static_cast<std::vector<std::size_t>::difference_type>(params.number_of_sidbs),
+            numbers.end(),
+            [this, &all_combinations](const auto begin, const auto end)
+            {
+                std::vector<std::size_t> combination{};
+                combination.reserve(params.number_of_sidbs);
+
+                for (auto it = begin; it != end; ++it)
+                {
+                    combination.push_back(static_cast<std::size_t>(*it));
+                }
+
+                all_combinations.push_back(combination);
+
+                return false;  // keep looping
+            });
+
+        return all_combinations;
+    }
+    /**
+     * Checks if any SiDBs within the specified cell indices are located too closely together, with a distance of less
+     * than 0.5 nanometers.
+     *
+     * This function iterates through the provided cell indices and compares the distance between SiDBs. If it finds any
+     * pair of SiDBs within a distance of 0.5 nanometers, it returns `true` to indicate that SiDBs are too close;
+     * otherwise, it returns `false`.
+     *
+     * @param cell_indices A vector of cell indices to check for SiDB proximity.
+     * @return `true` if any SiDBs are too close; otherwise, `false`.
+     */
+    [[nodiscard]] bool are_sidbs_too_close(const std::vector<std::size_t>& cell_indices) noexcept
+    {
+        for (std::size_t i = 0; i < cell_indices.size(); i++)
+        {
+            for (std::size_t j = i + 1; j < cell_indices.size(); j++)
+            {
+                if (sidb_nanometer_distance<sidb_cell_clk_lyt_siqad>(skeleton_layout,
+                                                                     all_sidbs_in_cavas[cell_indices[i]],
+                                                                     all_sidbs_in_cavas[cell_indices[j]]) < 0.5)
+                {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    /**
+     * Add SiDB cells to a SiDB cell-level layout based on a vector of cell indices.
+     *
+     * This function creates a copy of the provided layout (`skeleton_layout`) and adds SiDB cells to it based on the
+     * cell indices provided in the `cell_indices` vector. If a cell at a specified index is of type
+     * `sidb_technology::cell_type::EMPTY`, it is replaced with `sidb_technology::cell_type::NORMAL` to represent the
+     * addition of SiDB cells.
+     *
+     * @param cell_indices A vector containing indices of cells in the layout to be added or modified.
+     * @return A copy of the original layout (`skeleton_layout`) with SiDB cells added at specified indices.
+     */
+    [[nodiscard]] Lyt add_cells_to_layout_based_on_indices(const std::vector<std::size_t>& cell_indices) noexcept
```

Comment:
From the function name, it is not really clear why a copy of a layout is returned. Also, in the docstring, it says "This function creates a copy of the provided layout", but the function doesn't receive any parameters.

Reply:
@marcelwa Many thanks for your commits! 

That's a good point. Thank you